### ISSUE TITLE
data_browser broken in 1.4

### ISSUE THREAD

**Issue by ercius** [ScopeFoundry/ScopeFoundry#47]
The relative imports for the `data_browser` added in [this commit](https://github.com/ScopeFoundry/ScopeFoundry/commit/f0194135abd3337482c5bf2a34bf3b5599629194) breaks the data browser.

```
>>> import ScopeFoundry.data_browser
CRITICAL:root:Unhandled exception:Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/opt/tljh/user/envs/TemDataBrowser_dev/lib/python3.10/site-packages/ScopeFoundry/data_browser/__init__.py", line 3, in <module>
    from .data_browser import DataBrowser
  File "/opt/tljh/user/envs/TemDataBrowser_dev/lib/python3.10/site-packages/ScopeFoundry/data_browser/data_browser.py", line 9, in <module>
    from .viewers.file_info import FileInfoView
ModuleNotFoundError: No module named 'ScopeFoundry.data_browser.viewers'
```

I think this was previously fixed in [this commit](https://github.com/ScopeFoundry/ScopeFoundry/commit/67534913bee51663593ca7c5b9c63598a53a8213) but has been reverted. I think having a package named `data_browser` with an `__init__` and a file inside that package with the same name (`data_browser.py`) should be avoided.

**Comment by saahitmog** [comment]
Is there a temporary solution to the issue without downgrading back to 1.3.0?

**Comment by ercius** [comment]
Last I hear the fix is still being debated. They might change `data_browser.py` to a different file name or move all of that code back into the `__init__.py`.

**Comment by UBene** [comment]
I think I identified the problem: If you pip install ScopeFoundry the subdirectrories of ScopeFoundry.data_browser are missing. 
@edbarnard this is probably due to a faulty config in pip-packaging.

@ercius In the meantime your TEM databrowser actually works if you have the hole ScopeFoundry repo copied in your 'main' folder
<img width="614" alt="image" src="https://github.com/ScopeFoundry/ScopeFoundry/assets/39636345/9f1bd403-b1cc-46bd-b164-83a063669b05">

**Comment by ercius** [comment]
@UBene Thanks! Pinning to the previous version is my preference since nothing changed in the data browser except the folder/file architecture.

**Comment by edbarnard** [comment]
Is there a quick test procedure to check if any fixes work for this issue? I will try updating the setup.py script to see if its fixed.

**Comment by UBene** [comment]
I think the test would be as simple as 1. Pip install scopefoundry 2. Download TEM data browser and run it. Again the pip install should add “viewers” and “plugin” subdir to data browser (as found in the repository). On Jun 28, 2024, at 6:14 PM, Edward Barnard ***@***.***> wrote:﻿
Is there a quick test procedure to check if any fixes work for this issue? I will try updating the setup.py script to see if its fixed.

**Comment by ercius** [comment]
You could also try `pip install -e .` in the folder that contains setup.py. That installs it as editable locally. If that reproduces the error (can't import data_browser) then you can easily make changes and iterate on the correct solution.

**Comment by ercius** [comment]
In my opinion, this is not a packaging issue. There are relative imports and other issues with the folder structure.

I just created a new environment using conda. Then I did `pip install ScopeFoundry pyqtgraph PyQt6`. I then can find where ScopeFoundry is in the file system:

```python
$ python
>>> import ScopeFoundry
>>> ScopeFoundry
<module 'ScopeFoundry' from '/opt/tljh/user/envs/temp/lib/python3.10/site-packages/ScopeFoundry/__init__.py'>
```

Listing that location shows that `data_browser` is installed:
```bash
$ ls /opt/tljh/user/envs/temp/lib/python3.10/site-packages/ScopeFoundry/data_browser/
data_browser_plug_in.py   data_browser.ui       logo.png
data_browser.py           data_browser_view.py  __pycache__
data_browser_test_app.py  __init__.py           scopefoundry_logo2C_1024.png
```

The import still fails:
```python
>>> import ScopeFoundry.data_browser
CRITICAL:root:Unhandled exception:Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/opt/tljh/user/envs/temp/lib/python3.10/site-packages/ScopeFoundry/data_browser/__init__.py", line 3, in <module>
    from .data_browser import DataBrowser
  File "/opt/tljh/user/envs/temp/lib/python3.10/site-packages/ScopeFoundry/data_browser/data_browser.py", line 9, in <module>
    from .viewers.file_info import FileInfoView
ModuleNotFoundError: No module named 'ScopeFoundry.data_browser.viewers'
```

Finally, Im not able to try to run the data browser directly:
```bash
percius:/opt/tljh/user/envs/temp/lib/python3.10/site-packages/ScopeFoundry/data_browser$ python data_browser.py
python data_browser.py
WARNING:root:ScopeFoundry unable to import iPython console, using pyqtgraph.console instead. Error: No module named 'IPython'
pyqdarktheme unavailable: No module named 'qdarktheme'
CRITICAL:root:Unhandled exception:Traceback (most recent call last):
  File "/opt/tljh/user/envs/temp/lib/python3.10/site-packages/ScopeFoundry/data_browser/data_browser.py", line 9, in <module>
    from .viewers.file_info import FileInfoView
ImportError: attempted relative import with no known parent package
```

**Comment by edbarnard** [comment]
This reproduces
```
conda env remove -n sftest
conda create -n sftest python=3.11
conda activate sftest
conda install numpy h5py qtpy qtconsole pyqt pyqtgraph qtconsole
pip install xreload uuid7 pyqtdarktheme
pip install ScopeFoundry=1.4.0
```

Then:

```
(sftest) esbarnard@esbstudio ~ % python
Python 3.11.9 (main, Apr 19 2024, 11:44:45) [Clang 14.0.6 ] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import ScopeFoundry.data_browser

CRITICAL:root:Unhandled exception:Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/esbarnard/opt/anaconda3/envs/sftest/lib/python3.11/site-packages/ScopeFoundry/data_browser/__init__.py", line 3, in <module>
    from .data_browser import DataBrowser
  File "/Users/esbarnard/opt/anaconda3/envs/sftest/lib/python3.11/site-packages/ScopeFoundry/data_browser/data_browser.py", line 9, in <module>
    from .viewers.file_info import FileInfoView
ModuleNotFoundError: No module named 'ScopeFoundry.data_browser.viewers'
```

**Comment by edbarnard** [comment]
This problem seems only show up when bundling for distribution ( `pip install -e .` doesn't seem to trigger this)

A change to `setup.py` seems to fix:

```diff
-    packages=['ScopeFoundry', 'ScopeFoundry.scanning','ScopeFoundry.examples','ScopeFoundry.graphics', 'ScopeFoundry.data_browser' ],
+    packages=['ScopeFoundry', 'ScopeFoundry.scanning','ScopeFoundry.examples','ScopeFoundry.graphics', 'ScopeFoundry.data_browser', 'ScopeFoundry.data_browser.viewers' ],
```     

```
conda env remove -n sftest
conda create -n sftest python=3.11
conda activate sftest
conda install numpy h5py qtpy qtconsole pyqt pyqtgraph qtconsole
pip install xreload uuid7 pyqtdarktheme
```

```
$ cd ScopeFoundry
$ python setup.py sdist
(creates a tar.gz file in dist/ScopeFoundry-x.x.x.tar.gz)
mkdir ~/test
cp dist/ScopeFoundry-x.x.x.tar.gz ~/test
cd ~/test
tar zxvf ScopeFoundry-x.x.x.tar.gz
cd ScopeFoundry-x.x.x

python setup.py build
python setup.py install
```

I will make a new version with this change

**Comment by UBene** [comment]
probably you also want to add plug-ins:

  packages=['ScopeFoundry',
'ScopeFoundry.scanning','ScopeFoundry.examples','ScopeFoundry.graphics',
'ScopeFoundry.data_browser', 'ScopeFoundry.data_browser.viewers',
*'ScopeFoundry.data_browser.plug_ins'* ]




Am Mo., 1. Juli 2024 um 18:48 Uhr schrieb Edward Barnard <
***@***.***>:

**Comment by edbarnard** [comment]
I have release v1.4.1 that includes this fix. Please reply if this fix does not work for you!

**Comment by edbarnard** [comment]
Using @ercius  suggestion of a GitHub CI/CD Action to publish to PyPI, I've now released v1.4.2 that hopefully fixes this problem fully. https://pypi.org/project/ScopeFoundry/1.4.2/
